### ISSUE TITLE
Channel stuck in TRANSIENT_FAILURE and no DNS Refresh

### ISSUE THREAD

**Issue by hollinwilkins** [grpc/grpc#14097]
Please answer these questions before submitting your issue. 
 
### Should this be an issue in the gRPC issue tracker?
Yes, this is a bug with gRPC deployed on Kubernetes using headless services.
 
### What version of gRPC and what language are you using?
Ruby 2.4.3
gRPC 1.8.3
 
### What operating system (Linux, Windows, …) and version?
Debian Jessie (the Docker image)
Specifically this Ruby docker image: https://github.com/docker-library/ruby/blob/1d0fe6a9d5c75d8a114fb03016f528d68c613d80/2.4/jessie/slim/Dockerfile
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 Ruby 2.4.3
 
### What did you do?
1. Created a gRPC deployment with 2 replicas
2. Created a headless service that points to both replicas
3. Created a client that uses a round_robin LB policy and a DNS resolver for creating subchannels
4. Made a few requests to the backend server, no problems
5. Destroyed the backend servers, causing 2 new servers to take their place (with different IP addresses)
6. Tried to make requests to the 2 new server instances after they come back up, all channels are stuck in TRANSIENT_FAILURE

GRPC_VERBOSITY=DEBUG GRPC_TRACE=client_channel,connectivity_state,round_robin GRPC_DNS_RESOLVER=ares

Showed no further attempts to refresh the DNS entries after the initial connection.

DNS lookup provides the following values, with a TTL of 30 seconds:
```
nslookup -qa=A -debug account-headless
*** Invalid option: qa=A
Server:		10.55.240.10
Address:	10.55.240.10#53

------------
    QUESTIONS:
	account-headless.production.svc.cluster.local, type = A, class = IN
    ANSWERS:
    ->  account-headless.production.svc.cluster.local
	internet address = 10.52.8.189
	ttl = 30
    ->  account-headless.production.svc.cluster.local
	internet address = 10.52.9.52
	ttl = 30
    AUTHORITY RECORDS:
    ADDITIONAL RECORDS:
------------
Name:	account-headless.production.svc.cluster.local
Address: 10.52.8.189
Name:	account-headless.production.svc.cluster.local
Address: 10.52.9.52
```
 
### What did you expect to see?
 I would expect the connection state to go into a TRANSIENT_FAILURE and when a reconnect attempt fails, c-ares should refresh DNS entries after the TTL. All channels to stale IPs should be dropped in favor of IPs returned in the DNS refresh.
 
### What did you see instead?
The client never attempted to refresh DNS entries and never made it out of the CONNECTING -> TRANSIENT_FAILURE loop, event after waiting and attempting RPC calls long after the DNS TTL expired.
 
Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

Will be happy to provide anything necessary to figure this issue out, just ask and I will get it to you promptly!
 
### Anything else we should know about your project / environment?
Kubernetes 1.8.4.gke1
Headless services

**Comment by hollinwilkins** [comment]
CCing @ejona86 ... Thank you for the suggestion to file a bug report and the suggestions on what to look for in debugging this.

**Comment by dpratt** [comment]
I'm seeing the same thing with a python client - the logs are continually spammed with 

```
I0125 21:57:40.566431376      51 subchannel.cc:431]          Failed to connect to channel, retrying
D0125 21:57:40.566436418      51 connectivity_state.cc:169]  SET: 0x7f78b0003420 subchannel: TRANSIENT_FAILURE --> CONNECTING [state_change] error=(nil) "No Error"
D0125 21:57:40.566441084      51 connectivity_state.cc:195]  NOTIFY: 0x7f78b0003420 subchannel: 0x7f78ac032388
D0125 21:57:40.566608167      51 round_robin.cc:423]         [RR 0x7f78b0002660] connectivity changed for subchannel 0x7f78b0003350, subchannel_list 0x7f78b0002180: prev_state=TRANSIENT_FAILURE new_state=CONNECTING p->shutdown=0 sd->subchannel_list->shutting_down=0 error="No Error"
D0125 21:57:40.566619628      51 connectivity_state.cc:169]  SET: 0x7f78b0002698 round_robin: CONNECTING --> CONNECTING [rr_connecting] error=(nil) "No Error"
D0125 21:57:40.566625048      51 subchannel_list.cc:62]      [round_robin 0x7f78b0002660] subchannel list 0x7f78b0002180 index 0 of 2 (subchannel 0x7f78b0003350): requesting connectivity change notification
D0125 21:57:40.566633568      51 connectivity_state.cc:118]  CONWATCH: 0x7f78b0003420 subchannel: from CONNECTING [cur=CONNECTING] notify=0x7f78ac032388
D0125 21:57:40.587333445      32 connectivity_state.cc:169]  SET: 0x7f78b0004560 subchannel: CONNECTING --> TRANSIENT_FAILURE [connect_failed] error=0x7f78b8156f20 {"created":"@1516917460.587289165","description":"Connect Failed","file":"src/core/ext/filters/client_channel/subchannel.cc","file_line":678,"grpc_status":14,"referenced_errors":[{"created":"@1516917460.587247062","description":"Failed to connect to remote host: FD shutdown","file":"src/core/lib/iomgr/ev_poll_posix.cc","file_line":470,"os_error":"Timeout occurred","referenced_errors":[{"created":"@1516917460.587239262","description":"connect() timed out","file":"src/core/lib/iomgr/tcp_client_posix.cc","file_line":111}],"target_address":"ipv4:10.1.111.13:8000"}]}
D0125 21:57:40.587351316      32 connectivity_state.cc:195]  NOTIFY: 0x7f78b0004560 subchannel: 0x7f78b408c738
I0125 21:57:40.587358080      32 subchannel.cc:683]          Connect failed: {"created":"@1516917460.587247062","description":"Failed to connect to remote host: FD shutdown","file":"src/core/lib/iomgr/ev_poll_posix.cc","file_line":470,"os_error":"Timeout occurred","referenced_errors":[{"created":"@1516917460.587239262","description":"connect() timed out","file":"src/core/lib/iomgr/tcp_client_posix.cc","file_line":111}],"target_address":"ipv4:10.1.111.13:8000"}
D0125 21:57:40.587372202      32 round_robin.cc:423]         [RR 0x7f78b0002660] connectivity changed for subchannel 0x7f78b0004490, subchannel_list 0x7f78b0002180: prev_state=CONNECTING new_state=TRANSIENT_FAILURE p->shutdown=0 sd->subchannel_list->shutting_down=0 error={"created":"@1516917460.587289165","description":"Connect Failed","file":"src/core/ext/filters/client_channel/subchannel.cc","file_line":678,"grpc_status":14,"referenced_errors":[{"created":"@1516917460.587247062","description":"Failed to connect to remote host: FD shutdown","file":"src/core/lib/iomgr/ev_poll_posix.cc","file_line":470,"os_error":"Timeout occurred","referenced_errors":[{"created":"@1516917460.587239262","description":"connect() timed out","file":"src/core/lib/iomgr/tcp_client_posix.cc","file_line":111}],"target_address":"ipv4:10.1.111.13:8000"}]}
D0125 21:57:40.587395341      32 subchannel_list.cc:62]      [round_robin 0x7f78b0002660] subchannel list 0x7f78b0002180 index 1 of 2 (subchannel 0x7f78b0004490): requesting connectivity change notification
D0125 21:57:40.587402865      32 connectivity_state.cc:118]  CONWATCH: 0x7f78b0004560 subchannel: from TRANSIENT_FAILURE [cur=TRANSIENT_FAILURE] notify=0x7f78b80483d8
I0125 21:57:40.587407765      32 subchannel.cc:477]          Retry immediately
I0125 21:57:40.587422723      32 subchannel.cc:431]          Failed to connect to channel, retrying
D0125 21:57:40.587429310      32 connectivity_state.cc:169]  SET: 0x7f78b0004560 subchannel: TRANSIENT_FAILURE --> CONNECTING [state_change] error=(nil) "No Error"
D0125 21:57:40.587435075      32 connectivity_state.cc:195]  NOTIFY: 0x7f78b0004560 subchannel: 0x7f78b80483d8
D0125 21:57:40.587514382      32 round_robin.cc:423]         [RR 0x7f78b0002660] connectivity changed for subchannel 0x7f78b0004490, subchannel_list 0x7f78b0002180: prev_state=TRANSIENT_FAILURE new_state=CONNECTING p->shutdown=0 sd->subchannel_list->shutting_down=0 error="No Error"
D0125 21:57:40.587585858      32 connectivity_state.cc:169]  SET: 0x7f78b0002698 round_robin: CONNECTING --> CONNECTING [rr_connecting] error=(nil) "No Error"
D0125 21:57:40.587593775      32 subchannel_list.cc:62]      [round_robin 0x7f78b0002660] subchannel list 0x7f78b0002180 index 1 of 2 (subchannel 0x7f78b0004490): requesting connectivity change notification
D0125 21:57:40.587603170      32 connectivity_state.cc:118]  CONWATCH: 0x7f78b0004560 subchannel: from CONNECTING [cur=CONNECTING] notify=0x7f78b80483d8
```

Note the target address of 'ipv4:10.1.111.13:8000' - this is the old IP address of one of my pods, and even after about 30 minutes of retry, it does not re-resolve the DNS entry.

**Comment by dgquintas** [comment]
Hello everyone. This was a known issue at that point in time (~Dec 2017). In a very timely way, I have [a two day old pull request that fixes this very issue](https://github.com/grpc/grpc/pull/14170). It's still in review and needs a small follow up set of changes with an ETA of next week. Unfortunately, these changes won't make it to the imminent 1.9 release, but will definitely be available in 1.10.

We did try to backport a mitigation to this issue in 1.8, but after spending several hours yesterday, we gave up. It didn't seem worthwhile to hack up a mitigation when a much cleaner fix (that addresses several other issues) is already in the works.
